x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -121,6 +121,32 @@ public synchronized Store.MetadataSnapshot openSession(String sessionUUID, Index
         }
     }
 
+    public void ensureSessionShardIdConsistency(String sessionUUID, ShardId shardId) {
+        final RestoreSession restore = onGoingRestores.get(sessionUUID);
+        if (restore == null) {
+            logger.debug("could not get session [{}] because session not found", sessionUUID);
+            throw new IllegalArgumentException("session [" + sessionUUID + "] not found");
+        }
+        final ShardId sessionShardId = restore.indexShard.shardId();
+        if (false == sessionShardId.equals(shardId)) {
+            throw new IllegalArgumentException(
+                "session [" + sessionUUID + "] shardId [" + sessionShardId + "] does not match requested shardId [" + shardId + "]"
+            );
+        }
+    }
+
+    public void ensureFileNameIsKnownToSession(String sessionUUID, String fileName) throws IOException {
+        final RestoreSession restore = onGoingRestores.get(sessionUUID);
+        if (restore == null) {
+            logger.debug("could not get session [{}] because session not found", sessionUUID);
+            throw new IllegalArgumentException("session [" + sessionUUID + "] not found");
+        }
+        // Ensure no file system traversal is possible by only allow file names known to the restore session
+        if (false == restore.getMetadata().contains(fileName)) {
+            throw new IllegalArgumentException("invalid file name [" + fileName + "]");
+        }
+    }
```

Comment:
I am not sure whether these methods need to be sychronized. I will seek advices from the distributed team.

Reply:
It's been advised that synchronization are not necessary for these methods since they do not mutate state.